Add PackingPlanBuilder

billonahill opened the pull request:
Adds `PackingPlanBuilder` to fix #946. This reduces almost all container state management code (i.e., map/list manipulation) from the packing algorithms and leaves just the algorithm logic itself.

There are a few more edits I'd need to make (possibly in others PR) but wanted to share this to get feedback on the approach:
- Add checking that all required setters were called before adding/removing instances (DONE)
- `Packing.pack` and `Packing.repack` should throw `PackingException` instead of returning null. I'd do this in another PR.
- Some of the `PackingUtils` code is now only used by `PackingPlanBuilder`, so it could move there as internal utils.
- Some unit tests for just `PackingPlanBuilder`.

cc/ @avflor @ashvina @maosongfu

avflor commented on heron/packing/src/java/com/twitter/heron/packing/PackingPlanBuilder.java:
```diff
@@ -0,0 +1,170 @@
+  private Resource maxContainerResource;
+  private Map<String, Long> componentRamMap;
+  private int requestedContainerPadding;
+  private int numContainers;
```
It's a bit confusing that both the builder and the individual algorithms track the number of containers. Would it be easier to use it in one place only?

avflor commented on heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java:
```diff
@@ -183,78 +183,43 @@ public PackingPlan pack() {
+      try {
+        PackingPlanBuilder planBuilder = newPackingPlanBuilder(currentPackingPlan);
+        planBuilder.updateNumContainers(numContainers);
+        planBuilder = getResourceCompliantRRAllocation(
```
If get resource compliant allocation fails with an exception, then the packingbuilder will still have the updaed number of containers? Shouldn't we bring it back to the old value in the catch clause? It might not make a difference eventually since we update again at the next iteration but it guarantees consistency. Now it is a little tricky to me to check whether the numContainers in the builder is always the correct one.

billonahill replied:
I think you're describing a scenario where a container is added and we then catch a `ResourceExceedException` and nothing has been added so we have an empty container. [We dont't have a check for that case and we should but] What currently happens is that we add a container and a number of instnaces are added to it until it's full and _then_ we get a `ResourceExceedException`, so we're consistent and the new container is full and it's time to add another.

avflor replied:
Ok makes sense.

avflor commented on heron/packing/src/java/com/twitter/heron/packing/PackingPlanBuilder.java:
```diff
@@ -0,0 +1,170 @@
+    return this;
+  }
+
+  public PackingPlanBuilder updateNumContainers(int count) {
```
I think it looks better if we allocate/deallocate  containers when the user invokes the updateNumContainers command and not leave it to addInstance which invokes init.

billonahill replied:
I haven't yet documented this, but the design intent (which we can enforce) is that the builder is used as follows:

1. Call the setters to set required settings (e.g., resource settings)
2. Set the container size
3. Add/remove instances
  3a. If more containers are required go to step 2.
4. Build

As such, there are a few reasons why allocation/deallocation happens in add/remove instances.
- setting container size is required before adding/removing because the system is bounded by container size. (Actually we should support either bounding by container size or containerMaxResources. Currently we require bounding by both.)
- instances can be added and removed, followed by build without the need to call set container size between the two.

avflor replied:
I'm a little confused here. The command to add a container is the following: PackingUtils.allocateNewContainer(containers, containers.get(0).getCapacity(), this.requestedContainerPadding) when someone adds a new container. Why can't we do this in the updatecommand? It doesn't have any connection to an instance.

Overall, what makes it hard for me to verify the correctness of the code is that the numContainers in the builder can change without allocating the necessary containers right away. Wouldn't it be easier to enforce that by requiring that when the numContainers is updated in the builder, the List<Containers> is also updated in a "transaction".

billonahill replied:
We can't allocate containers in the `updateNumContainers` because `setRequestedContainerPadding` might not have been called yet and we need the container padding. Hence we call 'initContainers' before each add/remove and we get our internal data structures consistent.

My plan was to document and enforce that steps 1 and 2 in my sequence above are done (in any order) to set the relevant params, before adding/removing instances. If make the decision to change the internal behavior to init containers upon update, we'll need to enforce more restrictions on the user ordering of actions that are not necessary and wouldn't be caught until runtime.

billonahill commented:
I've addressed all the TODOs in the diff except for throwing PackingException unstead of returning null, which I'll do separately. I've also simplified methods that are now redundant and the PackingPlanBuilder implementation. Let me know if there are additional comments. This is ready to commit IMO.

kramasamy commented:
@avflor @ashvina - can you let us know if you have any comments?

ashvina commented:
LGTM 👍
